### PR TITLE
Update 'java' → 'java-library' Gradle plugin, update deps configurations.

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Documentation: https://docs.gradle.org/4.2.1/userguide/java_library_plugin.html

Reasoning:

- `api` instead of `compile` shows intention better, ie RS should be available for RxJava users
- `testImplementation` instead of `testCompile` [should hide dependencies from RxJava pom.xml](https://search.maven.org/remotecontent?filepath=io/reactivex/rxjava2/rxjava/2.1.5/rxjava-2.1.5.pom)
- If [RxJava 3.x will actually be split into multiple modules](https://github.com/ReactiveX/RxJava/issues/5621), `api/implementation` configurations might be handy

Issues:

- JMH Gradle Plugin doesn't seem to properly support `api`, so I had to manually declare RS for `jmh` configuration

@akarnokd can you please check if Eclipse works ok with this?

If not, we probably need to add:

```groovy
plugins.withType(EclipsePlugin) {
    project.eclipse.classpath.plusConfigurations += [ configurations.api ]
}
```

### CODE REVIEW

**Comment by akarnokd** [comment]
It does work, but does it generate a proper POM?

**Comment by artem-zinnatullin** [comment]
I tried to check that, the problem is that locally on both `2.x` and this PR branches `gradlew generatePomFileForMavenJavaPublication` results in `pom.xml` that only contains:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>io.reactivex.rxjava2</groupId>
  <artifactId>rxjava</artifactId>
  <version>2.2.0-SNAPSHOT</version>
  <dependencies>
    <dependency>
      <groupId>org.reactivestreams</groupId>
      <artifactId>reactive-streams</artifactId>
      <version>1.0.1</version>
      <scope>compile</scope>
    </dependency>
  </dependencies>
</project>
```

Which is pretty different from RxJava [2.1.5 pom published to maven central](https://search.maven.org/remotecontent?filepath=io/reactivex/rxjava2/rxjava/2.1.5/rxjava-2.1.5.pom), so I'm not sure I can compare them locally.

// I stopped Gradle daemon and cleaned the project between switching branches, just in case.

**Comment by akarnokd** [comment]
You should be able to trigger a proper POM creating via this command line:

```
gradlew -PreleaseMode=full -PbintrayUser="a" -PbintrayKey="b" -PsonatypeUsername="c" -PsonatypePassword="d" build --stacktrace
```

This will, of course, fail to upload anything due to bad user/pass.

**Comment by artem-zinnatullin** [comment]
Didn't notice `releaseMode=full` when was looking at Gradle configs, will try!

**Comment by artem-zinnatullin** [comment]
So I've checked pom, looks like Bintray plugins don't care about `api/implementation` which is okay for now. Pom is good.

<details><summary>Expand to view pom.xml</summary><p>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>io.reactivex.rxjava2</groupId>
  <artifactId>rxjava</artifactId>
  <version>2.2.0-SNAPSHOT</version>
  <name>RxJava</name>
  <description>Reactive Extensions for Java</description>
  <url>https://github.com/ReactiveX/RxJava</url>
  <licenses>
    <license>
      <name>The Apache Software License, Version 2.0</name>
      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
      <distribution>repo</distribution>
    </license>
  </licenses>
  <developers>
    <developer>
      <id>akarnokd</id>
      <name>David Karnok</name>
      <email>akarnokd@gmail.com</email>
    </developer>
  </developers>
  <scm>
    <connection>scm:git:git@github.com:ReactiveX/RxJava.git</connection>
    <developerConnection>scm:git:git@github.com:ReactiveX/RxJava.git</developerConnection>
    <url>scm:git:git@github.com:ReactiveX/RxJava.git</url>
  </scm>
  <issueManagement>
    <system>github</system>
    <url>https://github.com/ReactiveX/RxJava/issues</url>
  </issueManagement>
  <dependencies>
    <dependency>
      <groupId>org.reactivestreams</groupId>
      <artifactId>reactive-streams</artifactId>
      <version>1.0.1</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.mockito</groupId>
      <artifactId>mockito-core</artifactId>
      <version>2.1.0</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.reactivestreams</groupId>
      <artifactId>reactive-streams-tck</artifactId>
      <version>1.0.1</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.testng</groupId>
      <artifactId>testng</artifactId>
      <version>6.11</version>
      <scope>test</scope>
    </dependency>
  </dependencies>
</project>
```
</p>
</details>

**Comment by artem-zinnatullin** [comment]
But I think this PR exposes greater problem: we shouldn't rely on Bintray plugins to compose pom. 

```console
gradlew publishToMavenLocal
``` 

Should publish exact same pom as we do for releases/snapshots. So anyone could properly test changes locally and add patched RxJava as own project dependency by simply using `mavenLocal()` repository.

This can be addressed in separate PR of course.

**Comment by artem-zinnatullin** [comment]
@akarnokd is there anything you want me to update here? 

Or you'd prefer first party support of new dependency configurations in other plugins (JMH, Bintray) before merging this?

**Comment by akarnokd** [comment]
I'll merge this after 2.1.6 has been released.

**Review by vanniktech — APPROVED** [review]

**Pull request merged**
